File: 32.Longest_Valid_Parentheses.py
"""
# 32. Longest Valid Parentheses.

# Topic: String, Dynamic Programming, Stack.

# Task:
-----------
Given a string containing just the characters '(' and ')', return the length of the longest 
valid (well-formed) parentheses substring.

Example 1:
Input: s = "(()"
Output: 2
Explanation: The longest valid parentheses substring is "()".

Example 2:
Input: s = ")()())"
Output: 4
Explanation: The longest valid parentheses substring is "()()".

Example 3:
Input: s = ""
Output: 0
 

Constraints:
0 <= s.length <= 3 * 10^4
s[i] is '(', or ')'.


# Testcase:
------------
"(()"
")()())"
""

# Code:
------------
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        

"""
# Solution:
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        # Initialize max_length to keep track of the maximum length of a valid substring
        max_length = 0
        # Initialize a stack with -1. This is useful for handling the base case and calculating lengths
        stack = [-1]

        # Iterate through each character in the string
        for i in range(len(s)):
            # If the current character is an opening bracket, push its index onto the stack
            if s[i] == '(':
                stack.append(i)
            else:
                # Pop the top element from the stack
                stack.pop()
                # If the stack is empty after popping, push the current index onto the stack
                if not stack:
                    stack.append(i)
                else:
                    # Calculate the length of the current valid substring
                    # Update the max_length if this length is greater than the current max_length
                    max_length = max(max_length, i - stack[-1])

        # Return the maximum length of the valid parentheses substring found
        return max_length


# Description:
'''
To solve the problem of finding the longest valid parentheses in a string, we can use a stack-based approach. 
The idea is to maintain a stack that keeps track of the indices of the parentheses in the string. This approach
helps in keeping track of the matching pairs of parentheses and calculating the lengths of valid substrings.

Here is the step-by-step algorithm:
-----------------------------------------
1. Initialize an empty stack and push -1 onto it. This acts as a base for the first valid substring.

2. Iterate through each character in the string:

- If the current character is '(', push its index onto the stack.
- If the current character is ')':
- Pop the top element from the stack.
- If the stack becomes empty after popping, push the current index onto the stack. This means we're starting a 
  new potential valid substring.
- If the stack is not empty, calculate the length of the current valid substring by subtracting the current index
  with the top element of the stack. Update the maximum length if necessary.

3. Return the maximum length found.

This solution efficiently calculates the length of the longest valid parentheses substring by using a stack to 
keep track of indices and updating the maximum length accordingly. 
It works for all cases, including empty strings and strings with no valid parentheses.

'''

File: Cpp/32. Longest Valid Parentheses.cpp
// 32. Longest Valid Parentheses.       - HARD - 

// Topic: String, Dynamic Programming, Stack.


/*
### Task:
---
Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses 
substring.

Example 1:
Input: s = "(()"
Output: 2
Explanation: The longest valid parentheses substring is "()".

Example 2:
Input: s = ")()())"
Output: 4
Explanation: The longest valid parentheses substring is "()()".

Example 3:
Input: s = ""
Output: 0
 
Constraints:
0 <= s.length <= 3 * 10^4
s[i] is '(', or ')'.


### Testcase:
---
"(()"
")()())"
""

### Code:
---
class Solution {
public:
    int longestValidParentheses(string s) {
        
    }
};

*/
// Solution: --------------------------------------

class Solution {
public:
    int longestValidParentheses(string s) {
        int maxLen = 0; // To keep track of the maximum length of valid parentheses
        stack<int> indices; // Stack to store the indices of '(' and the last unmatched ')'
        int start = -1; // Start index of a valid substring

        for (int i = 0; i < s.length(); ++i) {
            if (s[i] == '(') {
                // Push the index of '(' onto the stack
                indices.push(i);
            } else {
                if (indices.empty()) {
                    // No matching '(' for this ')', update start index
                    start = i;
                } else {
                    // Pop the matching '(' index
                    indices.pop();
                    if (indices.empty()) {
                        // If stack is empty, calculate length from the start index
                        maxLen = max(maxLen, i - start);
                    } else {
                        // Calculate length using the top index in the stack
                        maxLen = max(maxLen, i - indices.top());
                    }
                }
            }
        }

        return maxLen;
    }
};


// Description: ===================================
/*
To solve the "Longest Valid Parentheses" problem, we can use a stack to keep track of the indices of '(' characters 
and the start of a potential valid substring. The stack will help us match parentheses and find the lengths of valid substrings.

Here's a step-by-step approach to implementing the `longestValidParentheses` function:

1. **Initialize a Stack**: The stack will hold the indices of '(' characters and the last unmatched ')' character.
2. **Initialize Variables**: We need variables to keep track of the maximum length of valid parentheses found so 
     far (`maxLen`) and the start index of a potential valid substring (`start`).
3. **Iterate through the String**: For each character in the string, we do the following:
    - If the character is '(', we push its index onto the stack.
    - If the character is ')', we pop from the stack. If the stack is empty, this ')' has no matching '(', so we update 
      the start position to the next index. If the stack is not empty, we found a matching pair, and we update `maxLen` 
      to the maximum of its current value and the current index minus the top of the stack (to get the length of the valid substring).

### Description:

- The stack is used to track the positions of '(' characters. When we encounter a ')', we attempt to pop a '(' from the stack to match it. 
  If the stack is empty, it means the ')' does not have a matching '(' before it, so we update `start` to `i`, marking a new potential 
   start for a valid substring.

- If the stack is not empty after popping, it means we've found a valid pair of parentheses. We then calculate the length of this valid 
  substring. If the stack is empty after popping, it indicates that the current valid substring extends from the `start` position to the 
  current position `i`, so we calculate the length accordingly.

- The use of `maxLen = max(maxLen, i - indices.top());` ensures we always have the maximum length of a valid substring found so far. 
  The top of the stack either represents the index of the last unmatched '(', which serves as the beginning of the current valid 
  substring, or the last unmatched ')', which marks the start of the next valid substring.

- This solution has a time complexity of O(n), where n is the length of the string, because we traverse the string once. The space 
  complexity is O(n) in the worst case when all characters are '(' and need to be stored in the stack.

*/
